### PR TITLE
eq-525 summary page improvements

### PR DESCRIPTION
**Pull request by weapdiv-david**
### What is the context of this PR?
1. Fixes a bug whereby if a user POSTed to the summary page, then they would automatically be taken to the thank-you page.  
2. Restructures the summary page rendering for easier customisation, simpler templates, and better performance with screen readers.
3. Adds new layout to summary page including sectioning and an 'edit' link for each answer.
4. Improvements to accessibility of summary.
### How to review
1. Checkout the branch
2. Run the tests
3. Complete several surveys and check answer formatting is as expected.
4. Test 'Edit' links
- [ ] Do all commits have sensible commit messages?
- [ ] Is all new code unit tested?
- [ ] Are there integration tests if needed?
- [ ] Is there sufficient logging?
- [ ] Is there documentation or is the code self-describing?
### Who worked on the PR

@weapdiv-david
@hamishtaplin

### CODE REVIEW

**Comment by hamishtaplin** [comment]
This is already a huge improvement— is there any scope for simplifying the templating a bit? Being able to do something like this would make it easier to reason about I think:

``` jinja
  <dl class="summary">
    {% for block in summary.blocks %}
      <dt class="summary__title saturn">{{block.title}}</dt>
      <dd class="summary__block">
        <dl class="summary__items">
        {% for item in block.items %}
          <dt class="summary__question summary__question--{{item.type}}">{{item.question}}</dt>
          <dd class="summary__answer">
            <a href="{{item.link}}">{{item.answer}}</a>
          </dd>
        {% endfor %}
        </dl>
      </dd>
    {% endfor %}
  </dl>
```

File: app/questionnaire/questionnaire_manager.py [hunk 1]
```diff
@@ -144,13 +144,14 @@ def update_state(self, item_id, user_input):
     def _append(self, node):
         if not self._first:
             self._first = node
+            self._current = node
```
Comment (collisdigital) [inline app/questionnaire/questionnaire_manager.py]: this appears to be a reversion of the change made to move this code outside the if/else statement from before.

File: app/templating/summary/block.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+        self.state = state
+        self.items = self.build_summary_items(self.state, self.schema)
+        self.title = self.schema.title
+
+    def build_summary_items(self, state, schema):
+        items = []
+
+        for question_state in state.questions:
+            if question_state.display_on_summary:
+                items.append(SummaryItem.create_item(schema.questionnaire.get_item_by_id(question_state.id), question_state))
```
Comment (collisdigital) [inline app/templating/summary/block.py]: Readability of this is poor, can it be split up?

File: app/templating/summary/item.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+from app.utilities.factory import Factory
+
+
+class Item(object):
+    def __init__(self, schema, state):
+        self.schema = schema
+        self.state = state
+        self.question = self.schema.title or self.schema.answers[0].label
+        self.link = self.schema.container.container.id + '#' + self.schema.id
+        if len(self.schema.answers) > 0:
```
Comment (collisdigital) [inline app/templating/summary/item.py]: Either a comment or move this to a function with a good descriptive name as not clear what these two lines are doing and why

File: app/templating/summary/item.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+            self.type = self.schema.answers[0].type.lower()
+
+    @staticmethod
+    def create_item(schema, state):
+        from app.templating.summary.items.date_range_summary import DateRangeSummaryItem
+        from app.templating.summary.items.general_summary import GeneralSummaryItem
+
+        factory = Factory()
```
Comment (collisdigital) [inline app/templating/summary/item.py]: This whole factory should be factored out into its own file and only done once

File: app/templating/summary/items/general_summary.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    def __init__(self, schema, state):
+        super().__init__(schema, state)
+        self.answer = self.prepare_answer()
+
+    def prepare_answer(self):
+        if len(self.schema.answers) > 1:
+            # Currently, the only multi-answer questions (apart from date ranges) are in the census test
+            return None
+        else:
+            answer_type = self.schema.answers[0].type.upper()
```
Comment (collisdigital) [inline app/templating/summary/items/general_summary.py:13]: This all needs to be pulled out into a factory and simplified.

File: app/templating/summary/items/general_summary.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                elif answer_type == 'RADIO':
+                    return self._prepare_radio_value(self.state.answers[0].value)
+                elif answer_type == 'CHECKBOX':
+                    return self._prepare_checkbox_values(self.state.answers[0].value)
+                else:
+                    return self.state.answers[0].value
+            else:
+                return None
+
+    def _prepare_currency_answer(self, value):
```
Comment (collisdigital) [inline app/templating/summary/items/general_summary.py:28]: all of these should be moved to their own objects and registered with the factory

File: tests/app/templating/test_summary_template_preprocessor.py [hunk 7]
```diff
@@ -33,28 +33,26 @@ def setUp(self):
...
-        self.assertTrue(block2 in blocks)
-
-        self.assertIsNotNone(render_data['meta'])
-
-        MetaDataTemplatePreprocessor._get_metadata = original_get_metadata_method
-
-
-
-
+    # def test_build_view_data(self):
```
Comment (collisdigital) [inline tests/app/templating/test_summary_template_preprocessor.py]: Tests need fixing or removing, but test coverage should be maintained or improved.

File: app/questionnaire/questionnaire_manager.py (hunk 1 above)
Comment (hamishtaplin) [reply app/questionnaire/questionnaire_manager.py]: that might be my fault, I had some merge conflicts

**Comment by LJBabbage** [comment]
Doesn't work with conditional display, I see a question in the summary  I haven't seen in the survey. The 'In that case you'll love this question!' appears regardless

**Comment by ajmaddaford** [comment]
Closing as this is superseded by the work @LJBabbage is doing

**Pull request closed without merging**